src/functions.php

```diff
@@ -132,58 +167,128 @@ function isdir(string $path): Promise
  * If the path does not exist the returned Promise will resolve
  * to FALSE and will not reject with an error.
  *
- * @param string $path An absolute file system path
- * @return \Amp\Promise<bool>
+ * @param string $path An absolute file system path.
+ * @return Promise<bool>
  */
 function isfile(string $path): Promise
 {
-    return filesystem()->isfile($path);
+    $deferred = new Deferred;
+
+    stat($path)->onResolve(function (?Throwable $error, ?array $result) use ($deferred): void {
+        if ($result !== null) {
+            $deferred->resolve((bool) ($result["mode"] & 0100000));
+        } else {
+            $deferred->resolve(false);
+        }
+    });
+
+    return $deferred->promise();
+}
+
+/**
+ * Does the specified path exist and is it a symlink?
+ *
+ * If the path does not exist the returned Promise will resolve
+ * to FALSE and will not reject with an error.
+ *
+ * @param string $path An absolute file system path.
+ * @return Promise<bool>
+ */
+function isSymlink(string $path): Promise
+{
+    $deferred = new Deferred;
+
+    lstat($path)->onResolve(function (?Throwable $error, ?array $result) use ($deferred): void {
+        if ($result !== null) {
+            $deferred->resolve((bool) ($result["mode"] & 0020000));
```

Comment:
Not entirely sure about this. According to [stat docs](https://www.php.net/manual/en/function.stat.php), link is `0120000`. But using that value here fails the tests.

Reply:
A `(bool)` cast here isn't enough, you need to check `($mask & $mode) === $mask`, see https://3v4l.org/mbCiU. Does that fix the test?

Reply:
It does! Thanks, I'm not used to bitmask handling. This is good to know.